include/deal.II/dofs/dof_accessor.templates.h

```diff
@@ -1479,6 +1480,72 @@ namespace internal
         Assert (next == dof_indices.end (), ExcInternalError ());
       }
 
+
+
+      template <typename InputVector, typename ForwardIterator>
+      static
+      void extract_subvector_to(const InputVector &values,
+                                const types::global_dof_index *cache,
+                                const types::global_dof_index *cache_end,
+                                ForwardIterator local_values_begin)
+      {
+        values.extract_subvector_to (cache, cache_end, local_values_begin);
+      }
+
+
+
+#ifdef DEAL_II_WITH_TRILINOS
+      static
+      std::vector<unsigned int> sort_indices(const std::vector<types::global_dof_index> &v)
+      {
+        // initialize original index locations
+        std::vector<unsigned int> idx(v.size());
+        std::iota(idx.begin(), idx.end(), 0);
+
+        // sort indices based on comparing values in v
+        std::sort(idx.begin(), idx.end(),
+                  [&v](unsigned int i1, unsigned int i2)
+        {
+          return v[i1] < v[i2];
+        });
+
+        return idx;
+      }
+
+
+
+      template <typename ForwardIterator>
+      static
+      void extract_subvector_to(const LinearAlgebra::EpetraWrappers::Vector &values,
+                                const types::global_dof_index *cache,
+                                const types::global_dof_index *cache_end,
+                                ForwardIterator local_values_begin)
+      {
+        types::global_dof_index max = 0;
+        const types::global_dof_index *cache_begin = cache;
+        std::vector<types::global_dof_index> cache_indices;
+        const unsigned int cache_size = cache_end-cache_begin;
+        cache_indices.reserve(cache_size);
+        while (cache != cache_end)
+          {
+            if (*cache > max)
+              max = *cache;
+            cache_indices.push_back(*cache);
+            ++cache;
+          }
+
+        IndexSet index_set(max+1);
+        index_set.add_indices(cache_begin, cache_end);
+        index_set.compress();
+        LinearAlgebra::ReadWriteVector<double> read_write_vector(index_set);
+        read_write_vector.import(values, VectorOperation::insert);
+
+        // Copy the elements from read_write_vector and reorder them.
+        std::vector<unsigned int> sorted_indices_pos = sort_indices(cache_indices);
```

Comment:
Since you are sorting the indices already anyway, do you think it would make sense to create the sort indices as the first step, then create `cache_indices` as sorted list and add that sorted list to `index_set.add_indices`? Besides being faster inside the index set, I think it would make things a bit simpler, as the maximum computation would be simple.